### PR TITLE
Add `wp_theme_has_theme_json` as a public API to know whether a theme has a `theme.json`

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/45171

## What?

Adds a new public method called `wp_theme_has_theme_json()` that returns whether a theme has a `theme.json` or not.

## Why?

We don't have a public method for this common operation, and consumers have to resort to using private APIs such as `WP_Theme_JSON_Resolver::theme_has_support()`. This PR graduates the existing logic into a public method.

## TODO

- [x] deprecate `WP_Theme_JSON_Resolver::theme_has_support` https://github.com/WordPress/gutenberg/pull/45380
- [x] Remove the resolver's affected tests
- [x] implement the remaining tests

## Testing Instructions

Make sure the tests pass.

### CODE REVIEW

File: phpunit/data/themedir1/default/style.css [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/*
```
Comment (oandregal) [inline phpunit/data/themedir1/default/style.css:1]: Took the stylesheet from the `default` theme at [core](https://github.com/WordPress/wordpress-develop/tree/trunk/tests/phpunit/data/themedir1/default), so when this is ported tests don't need to be updated and work the same.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	 * @return boolean
+	 */
+	function wp_theme_has_theme_json( $use_cache = true ) {
+		static $theme_has_support = null;
+
+		if ( false === $use_cache ) {
+			$theme_has_support = null;
+		}
+
+		if ( isset( $theme_has_support ) ) {
```
Comment (aristath) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php]: nitpick: Using a strict `null !==` check would be easier to understand
```suggestion
		if ( null !== $theme_has_support ) {
```

File: phpunit/wp-theme-json.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<?php
+/**
+ * Tests theme.json related public APIs.
+ *
+ * @package Gutenberg
+ */
+
+class WP_Theme_Json_Test extends WP_UnitTestCase {
```
Comment (oandregal) [inline phpunit/wp-theme-json.php]: These tests are not picked up:

- `npm run test:unit:php`: does not report any error (all these tests should fail)
- `npm run test:unit:php -- --group theme_json`: does not run any test (it should run one, the last of these)

Comment (Mamaduka) [reply phpunit/wp-theme-json.php]: Try `npm run test:unit:php -- --filter WP_Theme_Json_Test`

Comment (Mamaduka) [reply phpunit/wp-theme-json.php]: The filenames should have a `-test.php` suffix.

Comment (oandregal) [reply phpunit/wp-theme-json.php]: Adding the suffix, fixes it. Thanks!

PS: it looks like others with the same issue were committed, fix at https://github.com/WordPress/gutenberg/pull/45288

**Comment by oandregal** [comment]
>  deprecate WP_Theme_JSON_Resolver::theme_has_support

We can't do this here. This is no longer in Gutenberg so it needs to be done in core directly.

File: phpunit/class-wp-theme-json-resolver-test.php [hunk 4]
```diff
@@ -125,19 +125,6 @@ public function test_translations_are_applied() {
 		);
 	}
 
-	public function test_switching_themes_recalculates_data() {
```
Comment (oandregal) [inline phpunit/class-wp-theme-json-resolver-test.php:128]: This has been moved to `phpunit/wp-theme-json-test.php`

File: phpunit/wp-theme-json-test.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+	/**
+	 * Test that it reports correctly child themes that do not have a theme.json
+	 * and the parent does.
+	 *
+	 * @group theme_json
+	 *
+	 * @covers wp_theme_has_theme_json
+	 */
+	public function test_child_theme_has_not_theme_json_but_parent_has() {
```
Comment (oandregal) [inline phpunit/wp-theme-json-test.php:83]: There are a couple of test for which the template resolution does not work as expected, it reports being the same as the stylesheet.

- `block-theme-child` for `test_child_theme_has_theme_json` test. Resolved as expected:

```php
[26-Oct-2022 15:46:26 UTC] stylesheet /wordpress-phpunit/includes/../data/themedir1/block-theme-child
[26-Oct-2022 15:46:26 UTC] template   /wordpress-phpunit/includes/../data/themedir1/block-theme
```

- `block-theme-child-no-theme-json` for `test_child_theme_has_not_theme_json_but_parent_has`. Not resolved as expected and fails:

```php
[26-Oct-2022 15:45:47 UTC] stylesheet /var/www/html/wp-content/themes/block-theme-child-no-theme-json
[26-Oct-2022 15:45:47 UTC] template /var/www/html/wp-content/themes/block-theme-child-no-theme-json
```

- `default-child-no-theme-json` for `test_child_theme_has_not_theme_json_but_parent_has`. Not resolved as expected, but passes:

```php
[26-Oct-2022 15:47:04 UTC] stylesheet /var/www/html/wp-content/themes/default-child-no-theme-json
[26-Oct-2022 15:47:04 UTC] template   /var/www/html/wp-content/themes/default-child-no-theme-json
```

**Comment by oandregal** [comment]
I need to figure out why for some of the tests the template directory is picked up incorrectly. See https://github.com/WordPress/gutenberg/pull/45168/files#r1005869347

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	 * Whether a theme or its parent have a theme.json file.
+	 *
+	 * @param boolean $clear_cache Whether the cache should be cleared and theme support recomputed. Default is false.
+	 *
+	 * @return boolean
+	 */
+	function wp_theme_has_theme_json( $clear_cache = false ) {
+		static $theme_has_support = null;
+
+		if ( false === $clear_cache ) {
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php]:
```suggestion
		if ( true === $clear_cache ) {
```

We should check against the `true` here unless I'm missing something.

File: phpunit/wp-theme-json-test.php (hunk 5 above)
Comment (oandregal) [reply phpunit/wp-theme-json-test.php:83]: Fixed by explicitely set up the theme directory https://github.com/WordPress/gutenberg/pull/45168/commits/92db7ba7fe63b2e1562e4ce6e492ff5edb662e87

**Comment by oandregal** [comment]
@Mamaduka @aristath this is ready to land, all test have been fixed :)

**Review by Mamaduka — APPROVED** [review]
Thank you, @oandregal.

The code looks good, and I've not seen any errors when testing manually.

Do we have an ticket for deprecating `WP_Theme_JSON_Resolver::theme_has_support()` in core? We can easily miss this during the backport period for WP 6.2.

**Comment by oandregal** [comment]
> Do we have an ticket for deprecating WP_Theme_JSON_Resolver::theme_has_support() in core? We can easily miss this during the backport period for WP 6.2.

I'll follow up with this.

**Comment by oandregal** [comment]
@Mamaduka While going to create the ticket, I've figured out a partial way to do it that will serve as a strong reminder: https://github.com/WordPress/gutenberg/pull/45380

It's not complete because we still need to remove all its usage in WordPress core, which we cannot do here. Though I thought it'd be more strong a reminder than having a ticket :)

**Comment by spacedmonkey** [comment]
All these references in core need to be updated.

**Comment by oandregal** [comment]
This is being backported at https://github.com/WordPress/wordpress-develop/pull/3556

**Comment by geriux** [comment]
Hey there 👋 after some testing we were able to confirm that this [change](https://github.com/WordPress/gutenberg/pull/45168/files#diff-90ef3e892cfc85d07f602384e427bae8a7524e84917ed304edbcf46cc07b09f8) affected the endpoint used by the mobile editor, to fetch the block editor settings, more specifically the global styles data we use to render the theme's colors.

This works ok for self-hosted sites but not in other environments. I tested that the changes in this [follow-up PR](https://github.com/WordPress/gutenberg/pull/45543) fixes the problem for the mobile endpoint.

**Pull request merged**